docs/DeveloperGuide.md

```diff
@@ -222,13 +242,49 @@ The `commons` package contains classes used by multiple other components in the
 
 This section describes some noteworthy details on how certain features are implemented.
 
+### **Overview** <a name="overview"></a>
+
+All commands in TrackIt@NUS follow a similar execution flow.
+
+![Command Activity Diagram](images/CommandActivityDiagram.png)
+
+We have done this to improve maintainability. Making our code more uniform will:
+
+* Make testing easier 
+* Allow us to track down bugs faster
+* Allow us to better use the Single Level of Abstraction (SLAP) principle in our code
+
+In addition, this will make our code base much more organised, and hence make it easier for new developers to quickly
+ learn and contribute.
+ 
+Another design challenge was how to manage our predicates. TrackIt@NUS makes use of many different predicates to
+ allow users to view specific tasks, lessons, and contacts. For exmaple, users can view module-specific tasks
+ , contacts, and lessons. They can also view overdue tasks and future tasks (tasks where the deadline is more than a
+  week away). To manage all these predicates, we had 2 options:
+
+|  | Pros | Cons |
+| ---- | ----- | ------- |
+| **Option 1 (current choice):** Extract each the predicates into their own unique class | Increases code maintainability and testability. Now as a developer you exactly where to find each predicate. Makes use of the DRY principle and improves abstraction because you no longer need to interact with the actual lambda or test function, simply call the predicate. | Makes code more verbose as each predicate can simply be declared using a single lambda. |
+| **Option2:** Declare each predicate using a single lambda in the ModelManager class. No predicate will have a class. | Makes code shorter and simpler to read. No need to create a class when you can simply declare a predicate with a lambda. | Need to duplicate such code when using ModelStubs for testing. This will violate the DRY principle. |
```

Comment:
I think this section should have a dedicated section like "Code design consideration", and also we should add in a "Feature design consideration"

Like this: 
![image](https://user-images.githubusercontent.com/20136490/98386288-f3099800-208a-11eb-84b8-24cffb9e5c68.png)

Reply:
Done